_source/_posts/2022-10-01-spring-vault.md

```diff
@@ -0,0 +1,639 @@
+
+```bash
+vault token create -output-curl-string -policy=vault-demo-app-policy
+```
+
+```bash
+curl -X POST -H "X-Vault-Token: $(vault print token)" -H "X-Vault-Request: true" -d '{"policies":["vault-demo-app-policy"],"ttl":"0s","explicit_max_ttl":"0s","period":"0s","display_name":"","num_uses":0,"renewable":true,"type":"service","entity_alias":""}' http://127.0.0.1:8200/v1/auth/token/create
+```
+
+You are now ready to update the config server. In the `vault-config-server` project, edit `src/main/resource/application.yml` to add Vault as the config backend:
+
+```yml
+server:
+  port: 8888
+
+spring:
+  profiles:
+    active: vault
+  cloud:
+    config:
+      server:
+        vault:
+          host: 127.0.0.1
+          port: 8200
+          kvVersion: 2
+logging:
+  level:
+    root: TRACE
+```
+
+Note that the logging level is set to TRACE to see the interaction between the server and Vault. Restart the `vault-config-server`.
+
+```shell
+./mvnw spring-boot:run
+```
+
+You should see the logs below if the server was configured correctly:
+
+```
+2020-05-01 01:19:10.105  INFO 14072 --- [main] SpringBootConfigurationServerApplication:
+ Started SpringBootConfigurationServerApplication in 4.525 seconds (JVM running for 4.859)
+```
+
+You will not see the config server trying to connect to Vault in the logs. That will happen when a config client requests the properties. Start the `vault-demo-app`, passing the token just created in the environment variable `SPRING_CLOUD_CONFIG_TOKEN`. For example:
+
+```shell
+SPRING_CLOUD_CONFIG_TOKEN=hvs.CAESIKd9pYyc9xesiqmwvep... \
+./mvnw spring-boot:run
+```
+
+When the `vault-demo-app` starts, it will request the configuration to the config server, which in turn will make a REST to Vault. In the config server logs, with enough logging level, you will be able to see:
+
+```shell
+2022-09-09 19:21:57.778 DEBUG 12359 --- [nio-8888-exec-1] ...: Received [GET /vault-demo-app/dev HTTP/1.1
+Accept: application/vnd.spring-cloud.config-server.v2+json
+X-Config-Token: hvs.CAESIKd9pYyc9xesiqmwvep...
+User-Agent: Java/17
+Host: localhost:8888
+Connection: keep-alive
+
+]
+```
+
+Go to `http://localhost:8080` and verify that authentication with Okta works.
+
+Finally, let's seal Vault. Sealing allows you to lock Vault data to minimize damages when an intrusion is detected.  In the container command line, enter:
+
+```shell
+vault operator seal
+```
+
+Restart `vault-demo-app` and verify the configuration will not be retrieved as Vault is sealed. The `vault-config-server` logs should read:
+
+```
+503 Service Unavailable: "{"errors":["Vault is sealed"]}<EOL>"]
+```
+
+To unseal Vault, run:
+
+```shell
+vault operator unseal {unsealKey}
```

Comment:
I'm able to unseal successfully:

```
/ # vault operator unseal $MY_KEY
Key             Value
---             -----
Seal Type       shamir
Initialized     true
Sealed          false
Total Shares    1
Threshold       1
Version         1.11.3
Build Date      2022-08-26T10:27:10Z
Storage Type    inmem
Cluster Name    vault-cluster-b034a2f6
Cluster ID      b4b893b4-164d-83fd-c4da-45094c85a68f
HA Enabled      false
```

However, I still get the 500 above after restarting the config server and the demo app.

Reply:
I can't reproduce it. I only restart the vault-demo-app after unsealing and it gets the config. With the same SPRING_CLOUD_CONFIG_TOKEN.